src/main/scala/workshop3/02 functional-operators.md

```diff
@@ -0,0 +1,237 @@
+      count += 1
+  }
+  count
+}
+```
+
+For this the collections API has defined a number of operators. Most notable here are [`foldLeft`] and [`foldRight`]. Which one to choose 
+depends entirely on what you're doing. `foldLeft` aggregates the list from left to right, whereas `foldRight` aggregates the list from 
+right to left. To do so (in both cases), these operators take two parameters: 
+
+* a seed value 
+* an aggregator function
+
+The aggregator function is first applied to the seed value and the first (`foldLeft`) or last (`foldRight`) element of the list, then
+the result of that is applied to the next value, etc. The final result is the result of the last application of the aggregator.
+The following pictures, taken from Wikipedia depict this process. Not the execution in both cases proceeds from the bottom of the
+to the top.
+
+[`foldLeft`]: http://www.scala-lang.org/api/current/index.html#scala.collection.Traversable@foldLeft[B](z:B\)\(op:\(B,A\)=%3EB):B
+[`foldRight`]: http://www.scala-lang.org/api/current/index.html#scala.collection.Traversable@foldRight[B](z:B\)\(op:\(A,B\)=%3EB):B
+
+![foldLeft](https://upload.wikimedia.org/wikipedia/commons/5/5a/Left-fold-transformation.png)
+
+![foldRight](https://upload.wikimedia.org/wikipedia/commons/3/3e/Right-fold-transformation.png)
+
+Typical seed-function combinations are:
+
+ description                                | seed       | aggregator
+--------------------------------------------|------------|--------------------
+ the sum of a list of numbers               | `0`        | `+`
+ subtracting all numbers from the seed      | any number | `-`
+ the product of a list of numbers           | `1`        | `*`
+ flatten a list                             | empty list | `++` (concat lists)
+ turning a list of characters into a String | `""`       | `+`(String concat)
+ counting the number of elements in a list  | `0`        | `+ 1`
+
+In code:
+
+```scala
+List(1, 2, 3, 4).foldLeft(0)(_ + _)  // or (i, sum) => i + sum
+List(1, 2, 3, 4).foldRight(0)(_ + _) // or (sum, i) => sum + i
+
+List(1, 2, 3, 4).foldLeft(100)(_ - _)  // or (i, rest) => i - rest
+List(1, 2, 3, 4).foldRight(100)(_ - _) // or (rest, i) => rest - i
+
+List(1, 2, 3, 4).foldLeft(1)(_ * _)  // or (i, prod) => i * prod
+List(1, 2, 3, 4).foldRight(1)(_ * _) // or (prod, i) => prod * i
+
+List(List(1, 2, 3), List(4, 5, 6), List(7, 8, 9)).foldLeft(List[Int]())(_ ++ _)  // or (acc, list) => acc ++ list
+List(List(1, 2, 3), List(4, 5, 6), List(7, 8, 9)).foldRight(List[Int]())(_ ++ _) // (list, acc) => list ++ acc
+
+List('a', 'b', 'c', 'd').foldLeft("")(_ + _)  // or (str, c) => str + c
+List('a', 'b', 'c', 'd').foldRight("")(_ + _) // or (c, str) => c + str
+
+List('a', 'b', 'c', 'd').foldLeft(0)((count, _) => count + 1)
+List('a', 'b', 'c', 'd').foldRight(0)((_, count) => count + 1)
+```
+
+Notice that the result of subtraction is the same for `foldLeft` and `foldRight`, but execution follows a different path
```

Comment:
This paragraph about folding with subtraction was intended differently. For all the other examples above it doesn't matter whether you use `foldLeft` or `foldRight` in combination with `_ operator _`. However, `List(1,2,3,4).foldLeft(20)(_ - _) == 10` whereas `List(1,2,3,4).foldRight(20)(_ - _) == 18`. This is because subtraction is not associative. I see I made a mistake myself in my version as well by reversing `rest` and `i` in `val res = ...`. Could you revert your changes and look at it again with this explanation?